src/renderer/components/monaco-editor/monaco-editor.tsx

```diff
@@ -0,0 +1,290 @@
+  }
+
+  componentDidMount() {
+    try {
+      this.createEditor();
+      this.logger?.info(`[MONACO]: editor did mount`);
+    } catch (error) {
+      this.logger?.error(`[MONACO]: mounting failed: ${error}`, this);
+    }
+  }
+
+  componentWillUnmount() {
+    this.unmounting = true;
+    this.saveViewState();
+    this.destroy();
+  }
+
+  private createEditor() {
+    if (!this.containerElem || this.editor || this.unmounting) {
+      return;
+    }
+    const { language, theme, readOnly, value: defaultValue } = this.props;
+
+    this.editor = editor.create(this.containerElem, {
+      model: this.model,
+      detectIndentation: false, // allow `option.tabSize` to use custom number of spaces for [Tab]
+      value: defaultValue,
+      language,
+      theme,
+      readOnly,
+      ...this.options,
+    });
+
+    this.logger?.info(`[MONACO]: editor created for language=${language}, theme=${theme}`);
+    this.validateLazy(); // validate initial value
+    this.restoreViewState(); // restore previous state if any
+
+    if (this.props.autoFocus) {
+      this.editor.focus();
+    }
+
+    const onDidLayoutChangeDisposer = this.editor.onDidLayoutChange(layoutInfo => {
+      this.props.onDidLayoutChange?.(layoutInfo);
+    });
+
+    const onValueChangeDisposer = this.editor.onDidChangeModelContent(event => {
+      const value = this.editor.getValue();
+
+      this.props.onChange?.(value, event);
+      this.validateLazy(value);
+    });
+
+    const onContentSizeChangeDisposer = this.editor.onDidContentSizeChange((params) => {
+      this.props.onDidContentSizeChange?.(params);
+    });
+
+    this.dispose.push(
+      reaction(() => this.model, this.onModelChange),
+      reaction(() => this.props.theme, editor.setTheme),
+      reaction(() => this.props.value, value => this.setValue(value)),
+      reaction(() => this.options, opts => this.editor.updateOptions(opts)),
+
+      () => onDidLayoutChangeDisposer.dispose(),
+      () => onValueChangeDisposer.dispose(),
+      () => onContentSizeChangeDisposer.dispose(),
+      this.bindResizeObserver(),
+    );
+  }
+
+  destroy(): void {
+    if (!this.editor) return;
+
+    this.dispose();
+    this.editor.dispose();
+    this.editor = null;
+  }
+
+  @action
+  setDimensions(width: number, height: number) {
+    this.dimensions.width = width;
+    this.dimensions.height = height;
+    this.editor?.layout({ width, height });
+  }
+
+  setValue(value = ""): void {
+    if (value == this.getValue()) return;
+
+    this.editor.setValue(value);
+    this.validate(value);
+  }
+
+  getValue(opts?: { preserveBOM: boolean; lineEnding: string; }): string {
+    return this.editor?.getValue(opts) ?? "";
+  }
+
+  focus() {
+    this.editor?.focus();
+  }
+
+  @action
+  validate = (value = this.getValue()) => {
+    const validators: MonacoValidator[] = [
+      monacoValidators[this.props.language], // parsing syntax check
+    ].filter(Boolean);
```

Comment:
For future work, should we consider using the monaco language servers instead? That way we could get inline errors.

Reply:
Do you mean something from this list?

- https://langserver.org/
- https://github.com/TypeFox/monaco-languageclient

Adding this is a topic for another PR if we decide we need it.